framework/src/modules/chain/submodules/process_transactions.js

```diff
@@ -236,6 +263,75 @@ class ProcessTransactions {
 		// Add multisignature to transaction and process
 		return transaction.addMultisignature(stateStore, signature);
 	}
+
+	// eslint-disable-next-line class-methods-use-this
+	onBind(scope) {
+		library.modules = {
+			blocks: scope.modules.blocks,
+		};
+	}
+
+	/**
+	 * Executes each step from left to right and pipes the transactions that succeed to the next
+	 * step. Finally collects all responses and formats them accordingly.
+	 * @param steps
+	 * @returns {function(*=): {transactionsResponses: *[]}}
+	 */
+	static composeProcessTransactionSteps(...steps) {
+		return async transactions => {
+			let failedResponses = [];
+			const { transactionsResponses: successfulResponses } = await steps.reduce(
+				async (previousValue, fn, index) => {
+					if (index === 0) {
+						// previousValue === transactions argument in the first iteration
+						// First iteration includes raw transaction objects instead of formatted responses.
+						return fn(previousValue);
+					}
+
+					previousValue = await previousValue;
+
+					// Keep track of transactions that failed in the current step
+					failedResponses = [
+						...failedResponses,
+						...previousValue.transactionsResponses.filter(
+							response => response.status === TransactionStatus.FAIL
+						),
+					];
+
+					// Return only transactions that succeeded to the next step
+					return fn(
+						transactions.filter(transaction =>
+							previousValue.transactionsResponses
+								.filter(response => response.status === TransactionStatus.OK)
+								.map(transactionResponse => transactionResponse.id)
+								.includes(transaction.id)
+						)
+					);
+				},
+				transactions
+			);
+
+			return {
+				transactionsResponses: [...failedResponses, ...successfulResponses],
+			};
+		};
+	}
+
+	/**
+	 * Get current state from modules.blocks.lastBlock
+	 */
+	static _getCurrentContext() {
```

Comment:
We've changed to `static` because `this` looses context when `checkAllowedTransactions` is used with `composeProcessTransactionSteps`

Reply:
this is not used in this function though...?
Also, to maintain the scope, you can use lambda function

Reply:
I mean, when I try to use `this._getCurrentContext()` inside `checkAllowedTransactions` when using `composeTransactionSteps`, the `this` looses context.

Reply:
ah i see. I think this usage of `static` is wrong.
`this._getCurrentContext()` should not be allowed since it is not instance method.
It should be used as `ProcessTransaction._getCurrentContext()`
Im not sure if it applies to this case since the syntax is invalid, but in general if it looses the context, we can use lambda to insert or bind the context

Reply:
Binding will required a big code refactoring and will break the tests. Would be better to do this in another issue